src/backends/torch/torchlib.cc

```diff
@@ -534,15 +561,30 @@ namespace dd
     try
       {
         inputc.transform(ad);
-        if (_module._native)
+
+        _module._native = std::shared_ptr<NativeModule>(
+            NativeFactory::from_template(_template, _template_params, inputc));
+
+        if (_module._native != nullptr)
+          {
+            if (!this->_mlmodel._native.empty())
+              torch::load(_module._native, this->_mlmodel._native, _device);
+            _module._native->to(_device);
+          }
+
+        if (_module._graph)
           {
             auto batchoptional = inputc._dataset.get_batch({ 1 });
             TorchBatch batch = batchoptional.value();
             torch::Tensor td = batch.data[0];
-            _module._native->finalize(td.sizes());
+            _module._graph->finalize(td.sizes());
             // reload params after finalize
             if (!this->_mlmodel._traced.empty())
-              torch::load(_module._native, this->_mlmodel._traced, _device);
+              torch::load(_module._graph, this->_mlmodel._traced, _device);
+            _module._graph->to(_device);
+          }
+        if (_module._native)
+          {
             _module._native->to(_device);
```

Comment:
`module._native->to(_device)` called a second time (first time l.572)

It could be great to centralize calls of the `to()` method in `TorchModule`, unless we need to have some models (graph, native, traced) on cpu and some other on gpu. What do you think?

Reply:
ok, almost done :)